uk.co.compucorp.civicrm.hrcontactactionsmenu/CRM/HRContactActionsMenu/Page/UserMailNotifier.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+use CRM_HRCore_CMSData_UserMailNotifierFactory as CMSUserMailNotifierFactory;
+use CRM_HRContactActionsMenu_Helper_Contact as ContactHelper;
+
+/**
+ * Class CRM_HRContactActionsMenu_Page_UserMailNotifier
+ */
+class CRM_HRContactActionsMenu_Page_UserMailNotifier {
+
+  /**
+   * Function to send password reset email to CMS user
+   * of the civicrm contact.
+   */
+  public static function sendPasswordResetEmail() {
+    $contactID = CRM_Utils_Array::value('cid', $_GET);
+    $cmsUserMailNotifier = self::getCMSUserMailNotifier($contactID);
+    $cmsUserMailNotifier->sendPasswordResetEmail();
+
+    CRM_Core_Session::setStatus(ts('Password Reset Email sent'), 'Success', 'success');
+
+    $url = CRM_Utils_System::url('civicrm/contact/view', "reset=1&cid=$contactID");
+    CRM_Utils_System::redirect($url);
+  }
+
+  /**
+   * Function to send welcome email to CMS user
+   * of the civicrm contact.
+   */
+  public static function sendWelcomeEmail() {
+    $contactID = CRM_Utils_Array::value('cid', $_GET);
+    $cmsUserMailNotifier = self::getCMSUserMailNotifier($contactID);
+    $cmsUserMailNotifier->sendWelcomeEmail();
+
+    CRM_Core_Session::setStatus(ts('Welcome Email sent'), 'Success', 'success');
+
+    $url = CRM_Utils_System::url('civicrm/contact/view', "reset=1&cid=$contactID");
+    CRM_Utils_System::redirect($url);
+  }
+
+  /**
+   * Gets the CMSNotifier class depending on the CMS framework
+   *
+   * @param int $contactID
+   *
+   * @return \CRM_HRCore_CMSData_UserMailNotifierInterface
+   *
+   * @throws \Exception
+   */
+  private static function getCMSUserMailNotifier($contactID) {
+    $cmsFramework = CRM_Core_Config::singleton()->userFramework;
+    $contactUserInfo = ContactHelper::getUserInformation($contactID);
+    $contactUserInfo['cmsId'] = $contactUserInfo['id'];
+
+    return CMSUserMailNotifierFactory::create($cmsFramework, $contactUserInfo);
```

Comment:
If we call the `Civi::container()->get('hrcore.cms_notifier')->sendWelcomeEmail($user)`,
The problem would be to get the user object  as we would now be exposing the logic to get the user object to the contactactionsmenu extension and the extension does not know anything about the CMS user object.
We actually need only one identifier instance since the instance has the user information object. 
But for the case above, the url's will mostly be called one at a time anyway so a new instance would have to be created for each page callback.

Reply:
> object as we would now be exposing the logic to get the user object to the contactactionsmenu extension

What logic exactly are you talking about here? I think this is already already encapsulated in `ContactHelper::getUserInformation()`. Or are you talking about the `user_load()` call inside the `CRM_HRCore_CMSData_UserMailNotifier_Drupal` constructor? If it's about the latter, then we don't need to expose it, just move it out of the constructor to some other method in the same class.

Keep in mind that this notifier class can end up being used in other places (this is one of the reasons we moved it to the hrcore extension). Think about a scenario where we would need to notify multiple scenarios. With what I suggested, only a single instance of the notifier class would be necessary

Reply:
Yes, I'm referring to the user_load() call inside the CRM_HRCore_CMSData_UserMailNotifier_Drupal. 
Ok, will move it to a separate method then.
Also we should bear in mind that the other classes like Paths, Role in the CMS Data namespace all need one instance for each user, Are we planning to change them also?

Reply:
Yeah. Let's start with the `UserMailNotifier` and then we can change the other classes later. Tbh, I think we'll need to rethink this whole set of classes at some point, because there's some overlap between them and the `CRM_HRCore_Service_DrupalRoleService` and `CRM_HRCore_Service_DrupalUserService` classes